rustworkx-core/src/community/louvain.rs

```diff
@@ -0,0 +1,432 @@
+        return false;
+    }
+
+    // Compute the resulting new partition of the input graph
+    let input_graph = &partition.graph;
+    let mut final_index = HashMap::new();
+    let mut next_com = 0;
+    let mut updated_partition: Vec<usize> = vec![0; input_graph.node_count()];
+
+    for n in input_graph.node_identifiers() {
+        let prev_com = partition.subset_idx(n);
+        let inner_com = node_to_community[prev_com];
+        let new_com = if let Some(&c) = final_index.get(&inner_com) {
+            c
+        } else {
+            let c = next_com;
+            final_index.insert(inner_com, c);
+            next_com += 1;
+            c
+        };
+        updated_partition[input_graph.to_index(n)] = new_com;
+    }
+    partition.update(updated_partition);
+
+    true
+}
+
+/// Runs the Louvain community detection algorithm.
+///
+/// Arguments:
+///
+/// * `graph`: The input graph
+/// * `resolution` : controls whether the algorithm favors larger communities (`resolution < 1`) or smaller communities (`resolution < 1`)
+/// * `gain_threshold` : minimum acceptable increase in modularity at each level. The algorithm will
+///    terminate if it is not possible to meet this threshold by performing another level of aggregation.
+/// * `max_level` : Maximum number of levels (aggregation steps) to perform
+/// * `seed` : seed for RNG that determines the order in which we consider moving each node into a neighboring community
+pub fn louvain_communities<G>(
+    graph: G,
+    resolution: f64,
+    gain_threshold: f64,
+    max_level: Option<u32>,
+    seed: Option<u64>,
+) -> Vec<HashSet<G::NodeId>>
+where
+    G: Modularity,
+{
+    let mut partition = Partition::new(&graph);
+
+    let m = total_edge_weight(&graph);
+
+    let mut n_levels = 0;
+    while one_level_undirected(&mut partition, m, resolution, gain_threshold, seed) {
+        if let Some(limit) = max_level {
+            n_levels += 1;
+            if n_levels >= limit {
+                break;
+            }
+        }
+    }
+
+    partition.to_vec_of_hashsets()
+}
+
+#[cfg(test)]
+mod tests {
+    use crate::generators::barbell_graph;
+    use petgraph::graph::UnGraph;
+
+    use super::louvain_communities;
+
+    #[test]
+    fn test_louvain_barbell_graph() {
+        type G = UnGraph<(), f64>;
+
+        for n in 3..10 {
+            let g: G = barbell_graph(Some(n), Some(0), None, None, || (), || 1.0f64).unwrap();
+            let resolution = 1.0;
+            let gain_threshold = 0.01;
+            let result = louvain_communities(&g, resolution, gain_threshold, None, None);
+            // For a barbell graph, we expect the Louvain algorithm to identify
+            // the two complete subgraphs as the final communities
+            assert_eq!(result.len(), 2);
+            assert_eq!(result[0].len(), n);
+            assert_eq!(result[1].len(), n);
+        }
+    }
+
+    #[test]
+    fn test_louvain_karate_club_graph() {
```

Comment:
Friendly reminder to myself that we should add a Karate club graph generator. It could be a separate PR

Reply:
Once #1280 gets merged this test will hopefully be much shorter